src/v/cloud_storage/cache_service.cc

```diff
@@ -955,6 +955,49 @@ ss::future<> cache::stop() {
 }
 
 ss::future<std::optional<cache_item>> cache::get(std::filesystem::path key) {
+    std::vector<std::filesystem::path> keys = {key};
+    if (config::shard_local_cfg()
+          .cloud_storage_cache_num_buckets()
+          .has_value()) {
+        // If the config option was enabled and then disabled the objects will
+        // not be found in the cache and eventually be removed by cache
+        // eviction. Note that if the feature was disabled and then enabled the
+        // cache will be able to find both old objects and new objects. But if
+        // the feature was enabled and then disabled the objects in the cache
+        // will be inaccessible. They will be evicted and cache will be
+        // repopulated eventually.
+        auto rehashed = rehash_object_name(key);
+        if (rehashed.has_value()) {
+            vlog(
+              cst_log.debug,
+              "Get object name {} converted to {}",
+              key,
+              rehashed.value());
+            keys.emplace_back(std::move(rehashed.value()));
+        }
+    }
+    std::optional<cache_item> result;
+    for (auto k : keys) {
+        try {
+            result = co_await _get(k);
+        } catch (const std::filesystem::filesystem_error& e) {
+            if (e.code() != std::errc::no_such_file_or_directory) {
+                throw;
+            }
+        }
```

Comment:
in the lower level method we handle this code and return `std::nullopt`, so should we always throw here, and not handle exceptions at this level?